Update Mbed 5 boot sequence

c1728p9 opened the pull request:
### Description

Refactor the Mbed 5 boot process to make is simpler and more modular. This is done by breaking the boot sequence into 4 distinct steps - Target setup, Toolchain setup, RTOS setup and Mbed setup. This patch also move toolchain specific code into a per toolchain folder to make it more maintainable.

### Pull request type

    [ ] Fix
    [x] Refactor
    [ ] Target update
    [ ] Feature
    [ ] Breaking change

c1728p9 committed:
Refactor boot process

Refactor the Mbed 5 boot process to make is simpler and more modular.
This is done by breaking the boot sequence into 4 distinct steps -
Target setup, Toolchain setup, RTOS setup and Mbed setup. This patch
also move toolchain specific code into a per toolchain folder to make
it more maintainable.

c1728p9 committed:
Allow early use of singleton lock

Allow singleton_lock and singleton_unlock to be called before the
RTOS has been started by checking for a valid mutex before locking
and unlocking it.

c1728p9 commented:
Docs for this change can be found here:
https://github.com/ARMmbed/mbed-os-5-docs/pull/662

amq commented:
Is this related? https://github.com/ARMmbed/mbed-os/issues/7382

c1728p9 commented:
this is unrelated

bridadan commented on rtos/TARGET_CORTEX/mbed_boot.h:
```diff
@@ -0,0 +1,159 @@
+/**
+ * SDK hook for running code before ctors or OS
+ *
+ * This is a weak function which can be overridden at by a target's
```
English nit:

> This is a weak function which can be overridden **at by** a target's

should probably just be "by".

bridadan commented on rtos/TARGET_CORTEX/mbed_boot.h:
```diff
@@ -0,0 +1,159 @@
+/**
+ * Application hook for running code before main
+ *
+ * This is a weak function which can be overridden at by an application
```
English nit:

> This is a weak function which can be overridden **at by** an application

should probably just be "by".

bridadan reviewed (commented):
I realize I'm not on the reviewers list but I was perusing some of the PRs and noticed a couple of comment oopsies in this one! Very cool change though!

c1728p9 commented:
Thanks for the review @bridadan. I update the PR to fix those typos.

deepikabhavnani commented on platform/SingletonPtr.h:
```diff
@@ -43,6 +43,10 @@ extern osMutexId_t singleton_mutex_id;
 #ifdef MBED_CONF_RTOS_PRESENT
+    if (!singleton_mutex_id) {
+        // RTOS has not booted yet so no mutex is needed
+        return;
```
Should we throw an `MBED_ERROR` here?

c1728p9 replied:
nope, this case is expected to be hit early in the boot sequence when when the RTOS has not yet started.

hugueskamba replied:
It does seem to be the case for STM boards...

deepikabhavnani commented on rtos/TARGET_CORTEX/TOOLCHAIN_ARM/mbed_boot_arm.c:
```diff
@@ -0,0 +1,307 @@
+
+/* Define stack sizes if they haven't been set already */
+#if !defined(ISR_STACK_SIZE)
+    #define ISR_STACK_SIZE ((uint32_t)1024)
```
Define is used by ARM and GCC_ARM toolchain, can we move it to "mbed_boot.h" ?

c1728p9 replied:
done

deepikabhavnani reviewed (commented):
Thanks... this was much needed refactoring 👍

bulislaw commented on rtos/TARGET_CORTEX/TOOLCHAIN_GCC_ARM/mbed_boot_gcc_arm.c:
```diff
@@ -0,0 +1,125 @@
+
+void mbed_toolchain_init()
+{
+#if defined(MBED_CONF_RTOS_PRESENT)
```
I don't think we need this checks, nothing in `rtos/` folder will be compiled in if the RTOS is disabled.

c1728p9 replied:
done

bulislaw commented on rtos/TARGET_CORTEX/mbed_boot.h:
```diff
@@ -0,0 +1,159 @@
+/* mbed Microcontroller Library
```
You are using doxy, but it's not going to be anchored nicely. Is it worth creating a group and turning all the comments (even the file global one) into doxy?

c1728p9 replied:
done

bulislaw commented on rtos/TARGET_CORTEX/mbed_boot.h:
```diff
@@ -0,0 +1,159 @@
+ * - The RTOS has been started by a call to mbed_rtos_start
+ * - The toolchain has been initialized by a call to mbed_toolchain_init
+ */
+void mbed_main(void);
```
I'm guessing for compatibility reasons that need to stay, but from the conceptual point of view it doesn't have much sense as it's run from the same context than main. Would it make sense to add app entry point before RTOS and target entry point after RTOS?

c1728p9 replied:
It would be easy to add these hooks, although I would be hesitant to add them without having a use case in mind.

bulislaw commented on rtos/TARGET_CORTEX/mbed_rtos_rtx.c:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    osKernelStart();
+    while (1); // Code should never get here
```
Actually should we assert/error out here, rather than spin forever in case something goes wrong in RTX?

c1728p9 replied:
done

c1728p9 commented:
I made the following changes:
- Removed detailed boot procedure from mbed_boot.c since this is no longer relevent
- Added doxygen group to mbed_boot.h
- Added error if RTOS initialization fails
- Moved ISR_STACK_SIZE into mbed_boot.h
- Remove define MBED_CONF_RTOS_PRESENT

andrewleech commented:
I really like your goal and work you've done here. 
I've often found it incredibly hard to trace through the startup procedure, especially on new boards/targets when there's problems getting to main for any variety of reasons.

This is probably outside your desired scope, but would it take much effort to add an mbed config system entry to disable rtos aka: #7800

I regularly need to run mbed without the rtos on smaller targets that do not have enough ram/flash to support the os.

Currently the only way I know of to do this is adding the rtos and events libraries to `.mbedignore`
```
mbed-os/rtos/*
mbed-os/events/*
```
Unfortunately, I just tested this with this PR and it no longer works.

```
Compile [  0.3%]: main.cpp
[Fatal Error] Mutex.h@26,23: cmsis_os2.h: No such file or directory
[ERROR] In file included from ./mbed-os/features/netsocket/InternetSocket.h:25:0,
                 from ./mbed-os/features/netsocket/UDPSocket.h:23,
                 from ./mbed-os/features/netsocket/nsapi.h:40,
                 from ./mbed-os/mbed.h:26,
                 from .\main.cpp:1:
./mbed-os/rtos/Mutex.h:26:23: fatal error: cmsis_os2.h: No such file or directory
 #include "cmsis_os2.h"
                       ^
compilation terminated.
```

It would be great if you could make a better supported method of disabling the rtos!

andrewleech commented:
In case it wasn't clear in my last message,  
This PR breaks the ability to run mbed with rtos disabled. 

This may not be a particularly supported use case, but it is a critical need for many applications.

0xc0170 commented:
@andrewleech Thanks for the feedback. This is supported use case (this approach could be found in the issues/questions - its being used).

From your error log, what does `main.cpp` contain? The error comes from there. How to reproduce this failure? Looking at the changes here, I do not see where this would break it.

0xc0170 reviewed (approved):
Looks fine - nice cleanup! 

From the changes, this only touches rtos, not certain how this affects non rtos build as reported above. Please review it

andrewleech commented:
@0xc0170 It was a new project, fresh clone of mbed os with this branch checked out (0e7c3bfa) and the basic hello world main.cpp
```
#include "mbed.h"
 
DigitalOut myled(LED1);
 
int main() {
    while(1) {
        myled = 1;
        wait(0.2);
        myled = 0;
        wait(0.2);
    }
}
```
I tried compiling (I think for NRF52_DK) with `.mbedignore` as above and got that error. Stuck a '#' in front of both lines and it compiled fine.

Just tried it again, `mbed compile --clean -m NRF52_DK` with and without the lines commented, same result.

bridadan commented:
@0xc0170 The problem is probably the inclusion of rtos related headers in the toolchain boot source files. These rtos headers won't be present if the rtos directory is being ignored.

deepikabhavnani commented:
@andrewleech

I tried non-rtos build with blinky and shared .mbedignore and noted issue exists on the master branch as well. Our network interfaces are dependent on rtos, agree we have not done great job in separating those and adding appropriate errors. 

You can add a separate issue to mbed-os to take care of this in future. For now non-rtos blinky works with below:

```
mbed-os/rtos/*
mbed-os/events/*
mbed-os/features/*
```

0xc0170 commented:
Thanks @deepikabhavnani . Please review and leave feedback here. We can start CI soon

@andrewleech Please create an issue, we shall fix it !

andrewleech commented:
I'm so sorry I forgot to check master build as well, @c1728p9 sorry for the distraction. I love what you've done here.

NirSonnenschein commented:
conflicts need to be resolved

c1728p9 commented:
@NirSonnenschein I rebased this on top of master.

c1728p9 commented:
/morph build

cmonr commented:
The prep step got stuck somehow.

/morph build

0xc0170 commented:
I reviewed the failures, looks like all of them are the same `conflicting types for '__end__'` . Few platforms define it.

c1728p9 commented:
Updated this to fix `conflicting types for '__end__'`.

cmonr commented:
/morph build

cmonr commented:
/morph build

mprse commented on rtos/TARGET_CORTEX/TOOLCHAIN_GCC_ARM/mbed_boot_gcc_arm.c:
```diff
@@ -0,0 +1,115 @@
+/* Defined by linker script */
+extern uint32_t __end__[];
+#define HEAP_START      ((unsigned char*)__end__)
+#define HEAP_SIZE       ((uint32_t)((uint32_t)INITIAL_SP - (uint32_t)HEAP_START))
```
Why `INITIAL_SP` is used and not `ISR_STACK_START`.

I remember that `INITIAL_SP` is required for backward compatibility with mbed-rtos, but it would be good to not use it if possible, since `ISR_STACK_START` should define the same address.

mprse commented on rtos/TARGET_CORTEX/TOOLCHAIN_ARM/mbed_boot_arm.c:
```diff
@@ -0,0 +1,300 @@
+/* Defined by linker script */
+extern uint32_t Image$$RW_IRAM1$$ZI$$Limit[];
+#define HEAP_START      ((unsigned char*)Image$$RW_IRAM1$$ZI$$Limit)
+#define HEAP_SIZE       ((uint32_t)((uint32_t)INITIAL_SP - (uint32_t)HEAP_START))
```
Why `INITIAL_SP` is used and not `ISR_STACK_START`.

I remember that `INITIAL_SP` is required for backward compatibility with mbed-rtos, but it would be good to not use it if possible, since `ISR_STACK_START` should define the same address.

0xc0170 commented:
License server issue, resolved now, restarting

/morph export-build

0xc0170 commented:
/morph export-build

cmonr commented:
@c1728p9 Looks like this needs some more work.

c1728p9 commented:
Fixed ARM failures

cmonr commented:
/morph build

0xc0170 commented:
There's one device that hard faults, please verify

cmonr commented:
@c1728p9 Fyi ^^^

cmonr commented:
/morph build

cmonr commented:
Crud. Only meant to retest... Oh well.

cmonr commented:
GnuArmEclipse had an error initializing storage... Huh.
 /morph export-build

cmonr commented:
Reshuffling the Integration Test queue. Will restart shortly.

cmonr commented:
`Fatal error: A1023E: File "/tmp/filetQiJyd" could not be opened: No such file or directory`

This is less than useful...

/morph export-build

0xc0170 commented:
/morph export-build
/morph test

0xc0170 commented:
/morph test
